keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/ImageService.kt

```diff
@@ -83,110 +83,91 @@ class ImageService(
         }
       }
 
-  /**
-   * Get a specific image for an app version.
-   *
-   * @param region if supplied the latest image in this region is returned, if `null` the latest
-   * image regardless of region.
-   */
-  suspend fun getLatestNamedImage(appVersion: AppVersion, account: String, region: String? = null): NamedImage? =
-    cloudDriverService.namedImages(
-      user = DEFAULT_SERVICE_ACCOUNT,
-      imageName = appVersion.toImageName().replace("~", "_"),
-      account = account,
-      region = region
-    )
-      .filter { it.hasAppVersion }
-      .sortedWith(NamedImageComparator)
-      .firstOrNull {
-        // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
-        try {
-          AppVersion.parseName(it.appVersion).run {
-            packageName == appVersion.packageName && version == appVersion.version && commit == appVersion.commit
-          }
-        } catch (ex: Exception) {
-          throw SystemException("trying to parse name for version ${it.appVersion} but got an exception", ex)
+  private data class NamedImageCacheKey(
+    val appVersion: AppVersion,
+    val account: String,
+    val region: String
+  )
+
+  private val namedImageCache = cacheFactory
+    .asyncLoadingCache<NamedImageCacheKey, NamedImage>("namedImages") { (appVersion, account, region) ->
+      log.debug("Searching for baked image for {} in {}", appVersion.toImageName(), region)
+      cloudDriverService.namedImages(
+        user = DEFAULT_SERVICE_ACCOUNT,
+        imageName = appVersion.toImageName().replace("~", "_"),
+        account = account
+      )
+        // only consider images with tags and app version set properly
+        .asSequence()
+        .filter { image ->
+          tagsExistForAllAmis(image.tagsByImageId) && image.hasAppVersion
         }
-      }
+        // filter to images with matching app version
+        .filter { image ->
+          // TODO: Frigga and Rocket version parsing are not aligned. We should consolidate.
+          runCatching { AppVersion.parseName(image.appVersion) }
+            .getOrElse { ex ->
+              throw SystemException("trying to parse name for image ${image.imageName} with version ${image.appVersion} but got an exception", ex)
+            }
+            .run {
+              packageName == appVersion.packageName && version == appVersion.version && commit == appVersion.commit
+            }
+        }
+        // filter to images in the correct account and the desired region
+        .filter { image ->
+          image.accounts.contains(account) && image.amis.containsKey(region)
+        }
+        // reduce to the newest images required to support all regions we want
+        .sortedByDescending { it.creationDate }
+        .firstOrNull()
+  }
```

Comment:
This is the call to CloudDriver that backs the new cache. It looks for `appVersion` in the specified account and _single_ region. That filtering logic is the same as existed before but the region check is simpler now because it only cares about one.